### ISSUE TITLE
Add dedup run as a cron job on the server

### ISSUE THREAD

**Issue by loolmeh** [Tatoeba/tatoeba2#569]
I hope that we trust the script enough to let it run hourly. I'd need to rework the parsing of time intervals in the -i option to make it a bit more user friendly first though. Can I get a green light to go ahead with setting this up?

**Comment by jiru** [comment]
You want to make sure it plays well with other resource-consuming cron jobs like Sphinx reindexation (currently, delta every 15 mins and main every week). I’ve set up a lock thing so that concurrent reindexations shouldn’t happen (#553), but we may need something more global like a job queue that would encompass deduplication as well. The main problem is that this all put pressure the mysql daemon while as far as I know we can’t prioritize queries.

**Comment by tommy-3** [comment]
If you can make Horus work around the clock, it would be better if he could prevent us from adding a duplicate in the first place. That would keep the Corpus much neater.

The Horus ver.2 that I imagine would tell you when you try to add "Je t'aime." as a translation of "愛してる。",
"You cannot add the sentence 'Je t'aime.' because it already exists.
Click here to tag it OK.
Click here to link it to the Japanese sentence '愛してる。'"

Just an idea.

**Comment by Ppjet6** [comment]
On 2015/02/07, tommy-3 wrote:
> If you can make Horus work around the clock, it would be better if he could prevent us from adding a duplicate in the first place. That would keep the Corpus much neater.
> 
> The Horus ver.2 that I imagine would tell you when you try to add "Je t'aime." as a translation of "愛してる。",
> "You cannot add the sentence 'Je t'aime.' because it already exists.
> Click here to tag it OK.
> Click here to link it to the Japanese sentence '愛してる。'"
> 
> Just an idea.

I like this idea as well, but that's not something to be implemented as
a bot.

-- 
Maxime "Pepe_" Buquet

**Comment by trang** [comment]
Tommy brought up a point that I actually wanted to talk about. 

We definitely need a mechanism to detect duplicate when sentences are being added. I think for this we just could do a search with Sphinx every time someone adds a new sentence, and check in the results from Sphinx.

The deduplication script would then be needed only to merge duplicates that got through the initial check. In which case it perhaps won't be necessary to make Horus work every hour.

**Comment by jiru** [comment]
I agree with you, but I think the dedup script should be run periodically nevertheless. Duplicates can also be added by editing a sentence and I don’t think we want to prevent this from happening.

**Comment by loolmeh** [comment]
Making horus work every hour or 15 mins is already very cheap. Regarding handling this in the UI I'm not sure I can participate in that, probably needs another ticket but my thoughts on this is that there's probably 3 possible ways to implement this now that we don't have to worry about thousands of sentences that need to be merged first:

1- keep the current python script running at regular intervals, add a (text, lang) index in the db, then check directly against the database for duplicates manually and bubble up the result to the UI. [the easiest]

2- port core parts of the script to php, add a unique constraint for (text, lang), catch the exception in php then either bubble up as an error (for adding new sentences) or merge. [the hardest]

3- don't make any db changes, keep the current script, query sphinx to find a duplicate manually then bubble a message up to the UI. [easy but not reliable]

**Comment by trang** [comment]
Yes the topic of detecting duplicates when sentences are being added is not part of this issue. I've created another issue for it: #577.

As for the cronjob, I'm letting you guys decide how often you want to make Horus work. Anything's fine for me as long as you don't crash the server ^^

**Comment by Ppjet6** [comment]
Even if we use some temporary fix, in the end I would like to see lool0's 2. implemented. As discussed on the room, I know it's a lot of work, and that it's gonna break a lot of things, but I think it will be for the best, because here we're just implementing another layer on top of the db.

I'm sorry I'm just waking up, but I find it pretty impressive that Tatoeba made it until here without a unique constraint on (text, lang). There is also a 4th solution to be discussed (not viable yet) which is pytoeba.

**Comment by loolmeh** [comment]
I implemented a few improvements to the script for this ticket including a time parser for human readable inputs of relative time differences. The script now runs every 30 mins.

Notes:
- I tried adding an index on text, lang but not sure if it worked, innodb's antelope file format has a limit of 767 bytes on indices, nonetheless a single query takes ~6 ms on average, to put this into perspective, it took a few seconds to filter out, ~3k sentences from ~53k sentences from a time period of 2 weeks, the script itself took ~1hr to run.
- There's a master log at /var/log/dedup.log that's rotated weekly. At our current rate it would add ~7 mb of uncompressed data
- The script now runs every 30 mins and covers the previus 45 mins. We'd need to monitor it over the next week to see if we still need a run over the entire db, hopefully we won't need to. The command used takes around 0.4s to execute.
- script failures now don't kill the script and are logged in the log file.
